Sources/SwiftGenKit/Parsers/Strings/StringsDictEntry.swift

```diff
@@ -0,0 +1,216 @@
+    }
+  }
+
+  private static func decodeVariableRules(
+    in container: KeyedDecodingContainer<StringsDict.CodingKeys>,
+    formatKey: String
+  ) throws -> [PluralEntry.Variable] {
+    guard let variableKeyTuples = StringsDict.variableKeysFromFormatKey(formatKey) else { return [] }
+    let sortedVariableKeys = variableKeyTuples
+      .sorted { lhs, rhs in
+        // Sort by positional argument if present, otherwise by occurrence in the format key
+        switch (lhs.positionalArgument, rhs.positionalArgument, lhs.range.lowerBound, rhs.range.lowerBound) {
+        case let (.none, .none, lhs, rhs):
+          return lhs < rhs
+        case (.some, .none, _, _):
+          return true
+        case (.none, .some, _, _):
+          return false
+        case let (.some(lhs), .some(rhs), _, _):
+          return lhs < rhs
+        }
+      }
+      .map { $0.key }
+
+    return try sortedVariableKeys.reduce(into: [PluralEntry.Variable]()) { variables, variableKey in
+      let variableRule = try container.decode(PluralEntry.VariableRule.self, forKey: CodingKeys(key: variableKey))
+      variables.append(PluralEntry.Variable(key: variableKey, rule: variableRule))
+    }
+  }
+}
+
+// - MARK: Helpers
+
+extension StringsDict {
+  typealias VariableKeyResult = (key: String, range: Range<String.Index>, positionalArgument: Int?)
+
+  /// Parses variable keys and their ranges from a formatKey.
+  /// Every variable key that is contained within the format string is preceded
+  /// by the %#@ characters or by the positional variants, e.g. %1$#@, and followed by the @ character.
+  ///
+  /// - Parameter formatKey: The formatKey from which the variable keys should be parsed.
+  /// - Returns: An array of discovered variable keys and their range within the `formatKey`.
+  // swiftlint:disable:next discouraged_optional_collection
+  private static func variableKeysFromFormatKey(_ formatKey: String) -> [VariableKeyResult]? {
+    let pattern = #"(%(?>(\d)\$)?#@([\w\.\p{Pd}]+)@)"#
+    guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else { return nil }
+    let nsrange = NSRange(formatKey.startIndex..<formatKey.endIndex, in: formatKey)
+    let matches = regex.matches(in: formatKey, options: [], range: nsrange)
+
+    return matches.compactMap { match -> (String, Range<String.Index>, Int?)? in
+      // 1st capture group is the whole format string including delimeters
+      let rangeNSRange = match.range(at: 1)
+      // 2nd capture group is the positional argument of the format string (Optional!)
+      let positionalArgumentNSRange = match.range(at: 2)
+      // 3rd capture group is the key, the string in between the delimeters
+      let keyNSRange = match.range(at: 3)
+
+      guard rangeNSRange.location != NSNotFound, let rangeRange = Range(rangeNSRange, in: formatKey) else { return nil }
+      guard keyNSRange.location != NSNotFound, let keyRange = Range(keyNSRange, in: formatKey) else { return nil }
```

Comment:
Brain fart here, but doesn't `Range(nsRange, in: String)` already returns `nil` if `nsRange` is `NSNotFount`?

If so, that would allow you to simplify this code by getting rid of the check for `.location != NSNotFound` and only use the conditional binding to validate that (and same on line 170 below)

Reply:
Could be, would be nice if there were some documentation about `[Range(nsRange, in: String)](https://developer.apple.com/documentation/swift/range/3126733-init)`... I mainly used this kind of duplicated check, because it was used like that in the [NSHipster Regular Expressions](https://nshipster.com/swift-regular-expressions/#matching-multi-line-patterns-with-named-capture-groups) article.

Reply:
I just did a quick test in a Playground and it seems to work without the `NSNotFound` check.

Reply:
Turns out that this check is performed in the Range initializer itself:
https://github.com/apple/swift-corelibs-foundation/blob/master/Sources/Foundation/NSRange.swift#L286-L298

Reply:
Yeah I was gonna point to that. Sad that it's not documented in the official documentation page, as usual…